odoo/models.py

```diff
@@ -5831,19 +5831,19 @@ def modified(self, fnames, create=False, before=False):
                 records -= self.env.protected(field)
                 if not records:
                     continue
-                # Dont force the recomputation of compute fields which are
-                # not stored as this is not really necessary.
-                recursive = not create and field.recursive
+                marked_records = None
                 if field.compute and field.store:
-                    if recursive:
+                    if field.recursive:
                         marked_records = self.env.not_to_compute(field, records)
                     self.env.add_to_compute(field, records)
                 else:
-                    if recursive:
+                    # Dont force the recomputation of compute fields which are
+                    # not stored as this is not really necessary.
+                    if not create and field.recursive:
                         marked_records = records & self.env.cache.get_records(records, field)
                     self.env.cache.invalidate([(field, records._ids)])
                 # recursively trigger recomputation of field's dependents
-                if recursive:
+                if field.recursive and marked_records:
```

Comment:
The condition may be simplified to
```suggestion
                if marked_records:
```
You could even rename `marked_records` to `recursively_marked` to make things clearer.